### PR TITLE
[Collections] Signing (part 5): validate signature

### PR DESCRIPTION
**Pull request by yim-lee**
This continues [part 4](#3250) but doesn't necessarily depend on it.

Modifications:
- Add `PackageCollectionsSigning` dependency to `PackageCollections` module
- Verify collection signature in `JSONPackageCollectionProvider`

### CODE REVIEW

**Comment by yim-lee** [comment]
@swift-ci please smoke test

File: Sources/PackageCollectionsSigning/PackageCollectionSigning.swift [hunk 1]
```diff
@@ -15,7 +15,42 @@ import Basics
...
+    ///                     must be the first in the array.
+    ///   - certPolicyKey: The key of the `CertificatePolicy` to use for validating certificates
+    ///   - certPrivateKeyPath: Path to the private key (*.pem) of the certificate
+    ///   - jsonEncoder: The `JSONEncoder` to use
+    ///   - callback: The callback to invoke when the signed collection is available.
+    func sign(collection: PackageCollectionModel.V1.Collection,
+              certChainPaths: [URL],
+              certPrivateKeyPath: URL,
+              certPolicyKey: CertificatePolicyKey,
+              jsonEncoder: JSONEncoder,
```
Comment (tomerd) [inline Sources/PackageCollectionsSigning/PackageCollectionSigning.swift:33]: question: are we passing a concrete JSONEncoder to save the need to create a new one every time?

File: Sources/PackageCollectionsSigning/PackageCollectionSigning.swift [hunk 2]
```diff
@@ -15,7 +15,42 @@ import Basics
...
+    /// Validates a signed package collection.
+    ///
+    /// - Parameters:
+    ///   - signedCollection: The signed package collection
+    ///   - certPolicyKey: The key of the `CertificatePolicy` to use for validating certificates
+    ///   - jsonDecoder: The `JSONDecoder` to use
+    ///   - callback: The callback to invoke when the result is available.
+    func validate(signedCollection: PackageCollectionModel.V1.SignedCollection,
+                  certPolicyKey: CertificatePolicyKey,
+                  jsonDecoder: JSONDecoder,
```
Comment (tomerd) [inline Sources/PackageCollectionsSigning/PackageCollectionSigning.swift:47]: same question as above. seems a bit unnatural to have it in the public API vs. a  private static decoder/encoder at the implementation level (imo)

**Review by tomerd — APPROVED** [review]
reviewed the "part 5" commit (e23c7e7) only, looks good to me. one question regarding API design

File: Sources/PackageCollectionsSigning/PackageCollectionSigning.swift (hunk 1 above)
Comment (yim-lee) [reply Sources/PackageCollectionsSigning/PackageCollectionSigning.swift:33]: > are we passing a concrete JSONEncoder to save the need to create a new one every time?

No, the collection is serialized and embedded inside the signature, and upon checking the signature we deserialize the collection and compare it with the one received. In previous implementations we were comparing JSON/`Data` so we had to make sure we use the same encoder/decoder, but now that we are comparing `SignedCollection` we can probably get away with that.

**Comment by yim-lee** [comment]
Closing this PR, which has been replaced with #3271 and #3272

**Pull request closed without merging**